### PR TITLE
Enhancement Proposal EP00: Enhancement Proposals

### LINKED ISSUE #116: Document the Enhancement Proposal Workflow
Over in [a comment](https://github.com/FundingCircle/fc4-framework/pull/73#issuecomment-442224445) on #73 @sgerrand wrote:

> What's the workflow for proposals though? Do they get removed when the proposal is implemented? Could the content of a proposal be a GitHub issue instead? (So many questions. 😀)

I have a lot of thoughts on this, but it seems like a bit of a tangent from #73 itself, and it seems clear that we should document the EP workflow, hence this issue. I’ll post my initial thoughts as the top comment in a moment.

### PR DESCRIPTION
**Pull request by aviflax**
Fixes #116.

### CODE REVIEW

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+## Workflow
+
+This proposed workflow is a simplified version of that defined in [PEP 1](https://www.python.org/dev/peps/pep-0001/).
+
+### Starting an Enhancement Proposal (EP)
+
+An author should:
+
+1. Determine the ID number of the EP:
+   1. Check this GitHub repository’s for [open Pull Requests (PRs) with the label Enhancement
```
Comment (fussybeaver) [inline proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: `s/repository's/repository/`

Comment (fussybeaver) [inline proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Might want to quote "Enhancement Proposal" / italicise it to mark the exact label string.

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+      Proposal](https://github.com/FundingCircle/fc4-framework/pulls?q=is%3Aopen+is%3Apr+label%3A%22Enhancement+Proposal%22)
+      and if any are open, find the highest ID and increment that
+   1. If no EP PRs are open, the author can scan the contents of the directory `proposals` to find
```
Comment (fussybeaver) [inline proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Is it possible to have a link to the `proposals` directory here ?

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      the ID of the most recently published EP and increment that
+1. Create a new feature branch off of the latest master
+   1. The branch should be named `epXX-shortened-title`, e.g. `ep02-automated-rendering`
+1. Create a new directory under `proposals`, e.g. `proposals/ep02-automated-rendering`
+1. Create a new file in that directory, e.g.
+   `proposals/ep02-automated-rendering/ep02-automated-rendering.html`
+   1. Its initial contents should be those of the template below in [Appendix 1](#appendix-1), with
```
Comment (fussybeaver) [inline proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: This anchor bookmark did not work... I think you need `#appendix-1-ep-template`

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+      the $VARIABLES replaced with actual values
```
Comment (fussybeaver) [inline proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Consider using fixed-width backtick markdown for `$VARIABLES`

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+   1. It’s useful to do this now at this point, even though the EP isn’t
+      yet ready for review or discussion, in order to reserve the ID number
+   1. This may also help by making others aware that you’re working on this EP
+   1. Because the PR is not yet ready for review, include the prefix `WIP` in its title
+1. Add details to the EP, e.g. exactly what you’re proposing and why, the risks and tradeoffs, etc.
+
+### Discussing and Merging an EP PR
+
+1. Once the EP is ready for review and discussion, remove `WIP` from its title and request review
+   from relevant parties
```
Comment (fussybeaver) [inline proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Might need clarification here about who the relevant parties are, if they are specific people.

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+1. If a rough consensus develops around an EP that it’s cool, then:
```
Comment (fussybeaver) [inline proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: I personally would not describe an EP this way, when it is ready for an implementation pass. Perhaps just drop the "that it's cool".

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+   1. If it is likely to be implemented shortly then:
+      1. Its PR should be kept open until the implementation is merged
```
Comment (fussybeaver) [inline proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: `s/Its/The associated/` or similar.

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+      1. Just after the implementation is merged, its status should be changed to `final` and it
```
Comment (fussybeaver) [inline proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: "Its" is also confusing, do you mean the implementation's status or the EP's status ?

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+         should be merged
+   1. If it’s likely that the implementation will take awhile then:
+      1. Its status should be changed to `accepted` and it should be merged at that time, before the
+         implementation is completed
```
Comment (fussybeaver) [inline proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Ditto here, "Its" could mean the implementation or the EP.

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+      1. Its status should be changed to `final` in a subsequent follow-up PR, after the
+         implementation is completed
+1. If no rough consensus can be reached, or the rough consensus is that the proposed changes should
+   not be implemented, then the status should be changed to `withdrawn` and it should be merged.
```
Comment (fussybeaver) [inline proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Consider amending "the status" to "the status of the EP" to clarify what is being modified.

Comment (aviflax) [reply proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Addressed in c958153 — thanks!

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md (hunk 9 above)
Comment (aviflax) [reply proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Addressed in c958153 — thanks!

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md (hunk 8 above)
Comment (aviflax) [reply proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Addressed in c958153 — thanks!

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md (hunk 7 above)
Comment (aviflax) [reply proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Addressed in c958153 — thanks!

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md (hunk 6 above)
Comment (aviflax) [reply proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Addressed in c958153 — thanks!

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md (hunk 5 above)
Comment (aviflax) [reply proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Addressed in c958153 — thanks!

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md (hunk 4 above)
Comment (aviflax) [reply proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Addressed in c958153 — thanks!

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md (hunk 3 above)
Comment (aviflax) [reply proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Addressed in c958153 — thanks!

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md (hunk 2 above)
Comment (aviflax) [reply proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Addressed in c958153 — thanks!

File: proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md (hunk 1 above)
Comment (aviflax) [reply proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Addressed in c958153 — thanks!

Comment (aviflax) [reply proposals/ep00-enhancement-proposals/ep00-enhancement-proposals.md]: Addressed in c958153 — thanks!

**Pull request merged**